source/common/ssl/context_impl.cc

```diff
@@ -375,10 +425,41 @@ ServerContextImpl::ServerContextImpl(ContextManagerImpl& parent, const std::stri
         return dynamic_cast<ServerContextImpl*>(context_impl)->processClientHello(client_hello);
       });
 
-  if (!config.caCertFile().empty()) {
-    bssl::UniquePtr<STACK_OF(X509_NAME)> list(SSL_load_client_CA_file(config.caCertFile().c_str()));
-    if (nullptr == list) {
-      throw EnvoyException(fmt::format("Failed to load client CA file {}", config.caCertFile()));
+  if (!config.caCert().empty()) {
+    bssl::UniquePtr<BIO> bio(
+        BIO_new_mem_buf(const_cast<char*>(config.caCert().data()), config.caCert().size()));
+    RELEASE_ASSERT(bio != nullptr);
+    // Based on BoringSSL's SSL_add_file_cert_subjects_to_stack().
+    bssl::UniquePtr<STACK_OF(X509_NAME)> list(sk_X509_NAME_new(
+        [](const X509_NAME** a, const X509_NAME** b) -> int { return X509_NAME_cmp(*a, *b); }));
+    RELEASE_ASSERT(list != nullptr);
+    for (;;) {
+      bssl::UniquePtr<X509> cert(PEM_read_bio_X509(bio.get(), nullptr, nullptr, nullptr));
+      if (cert == nullptr) {
+        break;
```

Comment:
Do we have full test coverage for all these branches and exception cases?

Reply:
Unlikely, I'll add some.